crates/bevy_reflect/src/enums/enum_trait.rs

```diff
@@ -0,0 +1,280 @@
+/// valid names (such as `"3"`).
+/// This isn't wrong to do, but the convention set by the derive macro is that it isn't supported.
+/// It's preferred that these strings be converted to their proper `usize` representations and
+/// the [`Enum::field_at[_mut]`](Enum::field_at) methods be used instead.
+///
+/// [reflected]: crate
+/// [`None`]: core::option::Option<T>::None
+/// [`Some`]: core::option::Option<T>::Some
+/// [`Reflect`]: bevy_reflect_derive::Reflect
+pub trait Enum: Reflect {
+    /// Returns a reference to the value of the field (in the current variant) with the given name.
+    ///
+    /// For non-[`VariantType::Struct`] variants, this should return `None`.
+    fn field(&self, name: &str) -> Option<&dyn Reflect>;
+    /// Returns a reference to the value of the field (in the current variant) at the given index.
+    fn field_at(&self, index: usize) -> Option<&dyn Reflect>;
+    /// Returns a mutable reference to the value of the field (in the current variant) with the given name.
+    ///
+    /// For non-[`VariantType::Struct`] variants, this should return `None`.
+    fn field_mut(&mut self, name: &str) -> Option<&mut dyn Reflect>;
+    /// Returns a mutable reference to the value of the field (in the current variant) at the given index.
+    fn field_at_mut(&mut self, index: usize) -> Option<&mut dyn Reflect>;
+    /// Returns the index of the field (in the current variant) with the given name.
+    ///
+    /// For non-[`VariantType::Struct`] variants, this should return `None`.
+    fn index_of(&self, name: &str) -> Option<usize>;
+    /// Returns the name of the field (in the current variant) with the given index.
+    ///
+    /// For non-[`VariantType::Struct`] variants, this should return `None`.
+    fn name_at(&self, index: usize) -> Option<&str>;
+    /// Returns an iterator over the values of the current variant's fields.
+    fn iter_fields(&self) -> VariantFieldIter;
+    /// Returns the number of fields in the current variant.
+    fn field_len(&self) -> usize;
+    /// The name of the current variant.
+    fn variant_name(&self) -> &str;
+    /// The type of the current variant.
+    fn variant_type(&self) -> VariantType;
+    // Clones the enum into a [`DynamicEnum`].
+    fn clone_dynamic(&self) -> DynamicEnum;
+    /// Returns true if the current variant's type matches the given one.
+    fn is_variant(&self, variant_type: VariantType) -> bool {
+        self.variant_type() == variant_type
+    }
+    /// Returns the full path to the current variant.
+    fn variant_path(&self) -> String {
+        format!("{}::{}", self.type_name(), self.variant_name())
+    }
+}
+
+/// A container for compile-time enum info, used by [`TypeInfo`](crate::TypeInfo).
+#[derive(Clone, Debug)]
+pub struct EnumInfo {
+    type_name: &'static str,
+    type_id: TypeId,
+    variants: Box<[VariantInfo]>,
+    variant_indices: HashMap<Cow<'static, str>, usize>,
+}
+
+impl EnumInfo {
```

Comment:
Not something I feel inclined to block on, but we might want to add convenience `field` and `field_at` functions here, given that the Enum trait provides this pattern (much like Struct and StructInfo).

Reply:
Good idea, I'll make a note of it and add it in a followup PR!